pkg/app/pipedv1/livestatereporter/livestatereporter.go

```diff
@@ -0,0 +1,171 @@
+package livestatereporter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipecd/pkg/app/server/service/pipedservice"
+	"github.com/pipe-cd/pipecd/pkg/model"
+	pluginapi "github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1"
+	"github.com/pipe-cd/pipecd/pkg/plugin/api/v1alpha1/livestate"
+)
+
+type applicationLister interface {
+	ListByPluginName(name string) []*model.Application
+}
+
+type apiClient interface {
+	ReportApplicationLiveState(ctx context.Context, req *pipedservice.ReportApplicationLiveStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateResponse, error)
+	ReportApplicationSyncState(ctx context.Context, req *pipedservice.ReportApplicationSyncStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationSyncStateResponse, error)
+}
+
+// Reporter represents a component that reports the snapshot about live state of registered applications.
+type Reporter interface {
+	Run(ctx context.Context) error
+}
+
+type reporter struct {
+	reporters []pluginReporter
+	logger    *zap.Logger
+}
+
+// NewReporter creates a new reporter.
+func NewReporter(appLister applicationLister, apiClient apiClient, plugins map[string]pluginapi.PluginClient, logger *zap.Logger) Reporter {
+	rlogger := logger.Named("live-state-reporter")
+	r := &reporter{
+		reporters: make([]pluginReporter, 0, len(plugins)),
+		logger:    rlogger,
+	}
+
+	for name, p := range plugins {
+		r.reporters = append(r.reporters, pluginReporter{
+			pluginName:            name,
+			snapshotFlushInterval: time.Minute,
+			appLister:             appLister,
+			apiClient:             apiClient,
+			pluginClient:          p,
+			logger:                rlogger.With(zap.String("plugin-name", name)),
+		})
+	}
+	return r
+}
+
+func (r *reporter) Run(ctx context.Context) error {
+	group, ctx := errgroup.WithContext(ctx)
+
+	for i, reporter := range r.reporters {
+		// Avoid starting all reporters at the same time to reduce the API call burst.
+		time.Sleep(time.Duration(i) * 10 * time.Second)
+		r.logger.Info("starting app live state reporter for plugin")
+
+		group.Go(func() error {
+			return reporter.Run(ctx)
+		})
+	}
+
+	r.logger.Info(fmt.Sprintf("all live state reporters of %d plugins have been started", len(r.reporters)))
+
+	if err := group.Wait(); err != nil {
+		r.logger.Error("failed while running", zap.Error(err))
+		return err
+	}
+
+	r.logger.Info(fmt.Sprintf("all live state reporters of %d plugins have been stopped", len(r.reporters)))
+	return nil
+}
+
+type pluginReporter struct {
+	pluginName            string
+	snapshotFlushInterval time.Duration
+	appLister             applicationLister
+	apiClient             apiClient
+	pluginClient          pluginapi.PluginClient
+	logger                *zap.Logger
+}
+
+func (pr *pluginReporter) Run(ctx context.Context) error {
```

Comment:
Implemented it based on this code.
https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/livestatereporter/kubernetes/reporter.go#L79C1-L110C2